cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -306,4 +341,36 @@ private void publishStopProgram(String stopPayload) {
     programStateWriter.killed(programRunId);
     LOG.debug("Published program stop message for run {}", programRunInfo.getRun());
   }
+
+  /**
+   * Add notification to map if associated by a tethered peer
+   */
+  private Map<String, List<Notification>> getPeerProgramUpdates() {
+    Map<String, List<Notification>> peerToNotifications = new HashMap<>();
+    try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                    programUpdateTopic,
+                                                                    programUpdateFetchSize,
+                                                                    lastProgramUpdateMessageId)) {
+      if (iterator.hasNext()) {
+        Message message = iterator.next();
+        Notification notification = message.decodePayload(r -> GSON.fromJson(r, Notification.class));
+        if (notification.getNotificationType() == Notification.Type.PROGRAM_STATUS) {
+          Map<String, String> properties = notification.getProperties();
+          String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
+          try {
+            RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
+            if (runRecord.getPeerName() != null) {
+              peerToNotifications.getOrDefault(runRecord.getPeerName(), new ArrayList<>()).add(notification);
+            }
+          } catch (NotFoundException | IOException e) {
+            LOG.error("Unable to fetch runRecord for programRunId {}", programRunId, e);
+          }
+        }
+        lastProgramUpdateMessageId = message.getId();
```

Comment:
don't want to update this until after the POST call to the tethering service returns successfully, so that the messages will be re-read the next time.

Reply:
added a lastUnsentProgramUpdateMessageId to track this